MenuGuide.md (#218)
openmultiplayer/wiki: docs/tutorials/MenuGuide.md

> Creating a simple Menu

Reviewer: Please add metadata. Check this page out: https://www.open.mp/docs/meta/Contributing#metadata

======================

> ### From open.mp Wiki

Reviewer: Not necessary.

> Contents

Reviewer: Docusaurus generates a menu already, no need for this.

--------

*   1 Menus in Pawn
*   2 First menu steps
*   3 Add some menu items
*   4 Creating the effects for the items
*   5 Last steps
*   6 Last words
*   7 A full test script for menus

### 1 Menus in Pawn

> -------------

Reviewer: You only need one heading style, the accepted form is `#` characters, so the `------` is unnecessary.

> Menus look very complicated and difficult to script for the most players, although it isn't. Here I will show you how to create a simple menu. At the end we will have created a teleport menu.

Reviewer 2: ```suggestion
Menus look very complicated and difficult to script for the most players, although they're not. Here I will show you how to create a simple menu. At the end we will have created a teleport menu.
```

### 2 First menu steps
----------------

> First we have to create a menu. The prefix before is Menu: this makes the variable the correct type \[there are different types for different uses such as Float: bool: Text3D: etc\] Insert this line after the #includes, #defines, forwards etc:

Reviewer: https://www.open.mp/docs/meta/Contributing#use-code-snippets-for-technical-references

> First we have to create a menu. The prefix before is Menu: this makes the variable the correct type \[there are different types for different uses such as Float: bool: Text3D: etc\] Insert this line after the #includes, #defines, forwards etc:

Reviewer 2: ```suggestion
First we have to create a menu. We prefix our variable with `Menu` to give it the correct type. Insert this line after the #includes, #defines, forwards etc:
```

> new Menu:teleportmenu;

Reviewer: All code blocks should be surrounded by ```

> Okay, we created the variable to store the menu. Now we have to create the menu and assign the variable we created to the menu. Type this into OnGameModeInit():

Reviewer 2: ```suggestion
Okay, we created the variable to store the menu. Now we have to create the menu and assign the variable we created to it. Type this into `OnGameModeInit():`
```

teleportmenu = [CreateMenu](../scripting/functions/CreateMenu)("Teleportmenu", 2, 200.0, 100.0, 150.0, 150.0);

> Now a bit of an explanation about the CreateMenu arguments.

Reviewer 2: ```suggestion
Now a bit of an explanation about the CreateMenu arguments:
```

[CreateMenu](../scripting/functions/CreateMenu)(title, columns, Float:x, Float:y, Float:col1width, Float:col2width);
**Parameters:**

> |Parameter|Specifies|

Reviewer 2: ```suggestion
|Parameter|Description|
```

|--- |--- |

> |title|The heading of the menu

Reviewer 2: ```suggestion
|title|The heading of the menu|
```

> |columns|The number here defines how much columns are used \[2 is maximum\]|

Reviewer 2: ```suggestion
|columns|The number here defines how many columns are used \[2 is maximum\]|
```

> |Float:x|The heigth position of the menu on screen \[left to right\]|

Reviewer 2: ```suggestion
|Float:x|The `X` (horizontal) position of the menu on screen|
```

> |Float:y| The width position of the menu on screen \[up and down\]|

Reviewer 2: ```suggestion
|Float:y| The `Y` (vertical) position of the menu on screen |
```

|Float:col1width|The width of the first column|
|Float:col2width|The width of the second column|

### 3 Add some menu items
-------------------

> Ok, now we've got the Menu, but we need some items, under which you can choose in the menu. You add them underneath the CreateMenu that we made earlier.

Reviewer 2: ```suggestion
Ok, now we've got the Menu, but we need some items. You add them underneath the `CreateMenu` code that we made earlier.
```

> [AddMenuItem](../scripting/functions/AddMenuItem)(teleportmenu, 0, "LS");

Reviewer: This should be a code block and the URLs can be removed because they won't work in a code block anyway.

[AddMenuItem](../scripting/functions/AddMenuItem)(teleportmenu, 0, "LS");
[AddMenuItem](../scripting/functions/AddMenuItem)(teleportmenu, 0, "SF");
[AddMenuItem](../scripting/functions/AddMenuItem)(teleportmenu, 0, "SF");
[AddMenuItem](../scripting/functions/AddMenuItem)(teleportmenu, 0, "LV");
[AddMenuItem](../scripting/functions/AddMenuItem)(teleportmenu, 0, "LV");
 
[AddMenuItem](../scripting/functions/AddMenuItem)(teleportmenu, 1, "Grove Street");
[AddMenuItem](../scripting/functions/AddMenuItem)(teleportmenu, 1, "Starfish Tower");
[AddMenuItem](../scripting/functions/AddMenuItem)(teleportmenu, 1, "Wheel Arch Angels");
[AddMenuItem](../scripting/functions/AddMenuItem)(teleportmenu, 1, "Jizzys");
[AddMenuItem](../scripting/functions/AddMenuItem)(teleportmenu, 1, "4 Dragons");
[AddMenuItem](../scripting/functions/AddMenuItem)(teleportmenu, 1, "Come-a-Lot");

The explanation for AddMenuItem:

[AddMenuItem](../scripting/functions/AddMenuItem)(menuid, column, text);

> |menuid| The menuid of the menu where the item shall be displayed|
> |--|--|
> |column| The column in which the item shall be shown|
> |text|   The text of the item|

Reviewer 2: ```suggestion
|Parameter|Description|
|--|--|
|menuid| The menuid of the menu where the item shall be displayed|
|column| The column in which the item shall be shown|
|text| The text of the item|
```

### 4 Creating the effects for the items
----------------------------------

> Okay, now that we have created a full menu with items what should happen when you choose an item? In our example we want to make a teleportmenu, so we should get teleported to the position we choose. When a player selects an item on a menu the script calls the callback: public [OnPlayerSelectedMenuRow](../scripting/callbacks/OnPlayerSelectedMenuRow)(playerid, row). The best way to do it is to do it with a switch, this is like several if statements to check if a variable is worth certain values. But first we only want these effects for the menu we want so we need to create a variable that holds what menu the player is looking at, this is done with 'GetPlayerMenu':

Reviewer 2: ```suggestion
Okay, now that we have created a full menu with items what should happen when you choose an item? In our example we want to make a teleport menu, so we should get teleported to the position we choose. When a player selects an item on a menu the script calls the callback: public [OnPlayerSelectedMenuRow](../scripting/callbacks/OnPlayerSelectedMenuRow)(playerid, row). The best way to do it is to do it with a switch, this is like several if statements to check if a variable is equal to certain values. But first we only want these effects for the menu we want so we need to create a variable that holds what menu the player is looking at, this is done with `GetPlayerMenu`:
```

new Menu:CurrentMenu = [GetPlayerMenu](../scripting/functions/GetPlayerMenu)(playerid);

Now, when somebody selects something on the menu, their menuid will be saved in 'CurrentMenu'.

> Now we have to check that the menu they selected on is the menu we want:

Reviewer 2: ```suggestion
Now we have to check that the menu they selected is the menu we want:
```

public public [OnPlayerSelectedMenuRow](../scripting/functions/OnPlayerSelectedMenuRow)(playerid, row)
{
	new Menu:CurrentMenu = [GetPlayerMenu](../scripting/functions/GetPlayerMenu)(playerid);
	if (CurrentMenu == teleportmenu)
	{
    	//stuff
	}
	return 1;
}

> Now in between these brackets is where the 'switch' is, this checks what item the player selected or 'row' this can be done with if statements checking what row it is, but the switch is a much simpler way of writing it.

Reviewer 2: ```suggestion
Now in between these brackets is where the 'switch' is, this checks which item the player selected. This can be done with if statements, but the switch is a much simpler way of writing it.
```

        {
            [SetPlayerPos](../scripting/functions/SetPlayerPos)(playerid, 2028.5538, 1008.3543, 10.8203);
            [SetPlayerInterior](../scripting/functions/SetPlayerInterior)(playerid, 0);
            [SendClientMessage](../scripting/functions/SendClientMessage)(playerid, 0xFFFFFFFF, "Welcome to the Four Dragons Casino");
        }
        case 5: //Com-a-Lot
        {
            [SetPlayerPos](../scripting/functions/SetPlayerPos)(playerid, 2169.1838, 1122.5426, 12.6107);
            [SetPlayerInterior](../scripting/functions/SetPlayerInterior)(playerid, 0);
            [SendClientMessage](../scripting/functions/SendClientMessage)(playerid, 0xFFFFFFFF, "Welcome to the Come-a-Lot casino!");
        }
    }
}

### 5 Last steps
----------

Now we need a command to show the menu. This is the easiest step. Just a comparison with strcmp and a ShowMenuForPlayer. As you should know, this comes in the callback OnPlayerCommandText

if([strcmp](../scripting/functions/strcmp)(cmdtext, "/teleport", true) == 0)
{
    [ShowMenuForPlayer](../scripting/functions/ShowMenuForPlayer)(teleportmenu,playerid);
    return 1;
}

Really easy, wasn't it?

### 6 Last words
----------

> Okay, after you read this AND understood it, try your own menu. As you could see, it isn't that difficult, but will impress the players on your server all the more. And you can script really cool effects with this. It's also very cool for general stores or supermarkets for the things you can buy. Then you can subtract some money as effect and the price is shown in another column in the menu. But now, work on your own.

Reviewer 2: ```suggestion
Okay, after you've read and understood this, try your own menu. As you could see, it isn't that difficult, but will impress the players on your server. And you can script really cool effects with this. It's also very cool for general stores or supermarkets for the things you can buy. Then you can subtract some money as effect and the price is shown in another column in the menu.
```

> You can aslo add [TogglePlayerControllable](../scripting/functions/TogglePlayerControllable)(playerid,0); after ShowPlayerMenu(..);

Reviewer 2: ```suggestion
You can also add [TogglePlayerControllable](../scripting/functions/TogglePlayerControllable)(playerid,0); after ShowPlayerMenu(..);
```

> and  [TogglePlayerControllable](../scripting/functions/TogglePlayerControllable)(playerid,1) at end of OnPlayerSelectedMenuRow(..) so that player may not move while he is browsing menus.

Reviewer 2: ```suggestion
and  [TogglePlayerControllable](../scripting/functions/TogglePlayerControllable)(playerid,1) at the end of OnPlayerSelectedMenuRow(..) so that player may not move while they are browsing menus.
```
